### PR TITLE
miri engine: Hooks for basic stacked borrows

### PR DESCRIPTION
**Pull request by RalfJung**
r? @oli-obk

### CODE REVIEW

**Comment by RalfJung** [comment]
@bors r=oli-obk

(I had pushed something here but force-pushed it away again.)

**Comment by RalfJung** [comment]
@oli-obk I pushed some more commits, matching the [`stacked-borrows`](https://github.com/RalfJung/miri/commits/stacked-borrows) miri branch where the test suite all passes. :)

File: src/librustc_mir/interpret/cast.rs [hunk 1]
```diff
@@ -47,7 +46,19 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
 
             Misc => {
                 let src = self.read_value(src)?;
-                if self.type_is_fat_ptr(src_layout.ty) {
+
+                if src.layout.ty.is_region_ptr() && dest.layout.ty.is_unsafe_ptr() {
```
Comment (oli-obk) [inline src/librustc_mir/interpret/cast.rs]: slightly unhappy about code that isn't needed during const eval but still will be compiled into const eval. Not sure what to do about that though

Comment (RalfJung) [reply src/librustc_mir/interpret/cast.rs]: I cannot make it depend on `enforce_validity()` as that whitelists some functions in miri and I think my model might go crazy if references are only sometimes tracked.

I could add another `const REF_TRACKING_HOOKS: bool` though.

Comment (oli-obk) [reply src/librustc_mir/interpret/cast.rs]: Good idea, not a blocker for this PR though

**Comment by RalfJung** [comment]
Fixed your nit, and included https://github.com/solson/miri/pull/487 so that the miri submodule keeps working.

**Comment by RalfJung** [comment]
I fixed the conflict, and also added a [tiny change](https://github.com/rust-lang/rust/pull/55125/commits/964a20ddc205a4ede847c803e9471e7ffdf83c53) that bugged me after sleeping over this, and the corresponding miri update.

File: src/librustc_mir/interpret/cast.rs [hunk 2]
```diff
@@ -37,37 +37,34 @@ impl<'a, 'mir, 'tcx, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M>
...
 
             Misc => {
+                let src_layout = src.layout;
                 let src = self.read_value(src)?;
 
-                if M::ENABLE_PTR_TRACKING_HOOKS &&
-                    src.layout.ty.is_region_ptr() && dest.layout.ty.is_unsafe_ptr()
-                {
+                let src = if M::ENABLE_PTR_TRACKING_HOOKS && src_layout.ty.is_region_ptr() {
+                    assert!(dest.layout.ty.is_unsafe_ptr());
```
Comment (oli-obk) [inline src/librustc_mir/interpret/cast.rs]: add a comment mentioning that the only `Misc` casts that start out with a reference must be to a raw pointer. Clear in hindsight, but took me a minute to figure out why this assert is dead code barring bugs (which is exactly what it's trying to catch)

**Comment by oli-obk** [comment]
r=me with comment

**Comment by RalfJung** [comment]
Fixed by disabling a new miri test on Windows and macOS, where I cannot test them.

This can be much easier debugged on the miri side once things landed in rustc.

@bors r=oli-obk

**Pull request merged**